adapters/huaweiads/huaweiads.go

```diff
@@ -705,25 +789,31 @@ func convertCountryCode(country string) (out string) {
 }
 
 // getDeviceID include oaid gaid imei. In prebid mobile, use TargetingParams.addUserData("imei", "imei-test");
-func getDeviceID(device *device, openRTBRequest *openrtb2.BidRequest) (err error) {
-	if openRTBRequest.User == nil {
-		return errors.New("getDeviceID: openRTBRequest.User is nil.")
-	}
-	if openRTBRequest.User.Ext == nil {
-		return errors.New("getDeviceID: openRTBRequest.User.Ext is nil.")
+// When ifa: gaid exists, other device id can be passed by TargetingParams.addUserData("oaid", "oaid-test");
+// When getGaidFromDeviceIFA success, getDeviceIDFromUserExt failed, no error will be reported.
+func getDeviceIDFromUserExt(device *device, isGetGaidSuccess bool, openRTBRequest *openrtb2.BidRequest) (err error) {
+	if openRTBRequest.User == nil || openRTBRequest.User.Ext == nil {
+		if isGetGaidSuccess {
+			return nil
+		}
+		return errors.New("getGaidFromDeviceIFA failed, getDeviceIDFromUserExt failed: openRTBRequest.User.Ext is nil.")
 	}
 	var extUserDataHuaweiAds openrtb_ext.ExtUserDataHuaweiAds
 	if err := json.Unmarshal(openRTBRequest.User.Ext, &extUserDataHuaweiAds); err != nil {
-		return errors.New("Unmarshal: openRTBRequest.User.Ext -> extUserDataHuaweiAds failed")
+		if isGetGaidSuccess {
+			return nil
+		}
+		return errors.New("getGaidFromDeviceIFA failed, getDeviceIDFromUserExt failed: Unmarshal: openRTBRequest.User.Ext -> extUserDataHuaweiAds failed.")
 	}
 	var deviceId = extUserDataHuaweiAds.Data
-	if len(deviceId.Imei) == 0 && len(deviceId.Gaid) == 0 && len(deviceId.Oaid) == 0 {
-		return errors.New("getDeviceID: Imei ,Oaid, Gaid are all empty.")
+	if !isGetGaidSuccess && len(deviceId.Imei) == 0 && len(deviceId.Gaid) == 0 && len(deviceId.Oaid) == 0 {
+		return errors.New("getGaidFromDeviceIFA failed, getDeviceIDFromUserExt failed: Imei ,Oaid, Gaid are all empty.")
```

Comment:
I'd recommend not using the function names here in the error message.

Reply:
corrected